File: butler_pc_core/sidecar/routes/company_profile.py
from __future__ import annotations

import csv
from datetime import datetime, timezone
from io import StringIO
from typing import Any, Optional

from fastapi import APIRouter, File, Header, HTTPException, UploadFile
from pydantic import BaseModel, Field

from butler_pc_core.accounting.assignment.domain import AssignmentError
from butler_pc_core.auth.capability_token import (
    CapabilityTokenError,
    CapabilityTokenManager,
    auth_error_payload,
)
from butler_pc_core.company_policy.admin_auth import AdminAuthError, admin_error_payload, verify_admin_context
from butler_pc_core.company_policy.contracts import AdminContext
from butler_pc_core.company_profile.contracts import (
    ContractValidationError,
    make_runtime_profile,
    sha256_text,
)
from butler_pc_core.company_profile.normalizer import holder_alias_hint, normalize_holder_alias
from butler_pc_core.company_profile.storage import CompanyProfileStore, ProfileLoadError


router = APIRouter()
_STORE: CompanyProfileStore | None = None
_TOKEN_MANAGER = CapabilityTokenManager()


class OwnBankAccountRequest(BaseModel):
    bank_code: str = Field(..., min_length=2, max_length=16)
    account_number: str = Field(..., min_length=1, max_length=64)


class CompanyProfileRegisterRequest(BaseModel):
    own_bank_holder_aliases: list[str] = Field(..., min_length=1)
    own_company_aliases: list[str] = Field(..., min_length=1)
    own_account_numbers: list[str] = Field(default_factory=list)
    own_bank_accounts: list[OwnBankAccountRequest] = Field(default_factory=list)


def get_company_profile_store() -> CompanyProfileStore:
    global _STORE
    if _STORE is None:
        _STORE = CompanyProfileStore()
    return _STORE


def _admin_context(
    x_admin_role: Optional[str],
    x_admin_id_digest: Optional[str],
    x_admin_session_digest: Optional[str],
    x_admin_auth_method: Optional[str],
) -> AdminContext:
    if not x_admin_id_digest or not x_admin_session_digest:
        raise AdminAuthError("ADMIN_AUTH_REQUIRED", "admin digest headers required")
    return AdminContext(
        admin_id_digest=x_admin_id_digest,
        role=x_admin_role or "employee",  # type: ignore[arg-type]
        admin_session_digest=x_admin_session_digest,
        auth_method=x_admin_auth_method or "tauri_secure_invoke",  # type: ignore[arg-type]
    )


@router.get("/v1/company-profile/status")
async def company_profile_status(authorization: Optional[str] = Header(default=None)) -> dict[str, Any]:
    try:
        _TOKEN_MANAGER.verify_authorization_header(authorization)
    except CapabilityTokenError as exc:
        status = 401 if exc.fail_class.value == "CAPABILITY_TOKEN_MISSING" else 403
        raise HTTPException(status_code=status, detail=auth_error_payload(exc)) from exc

    try:
        entry = get_company_profile_store().load_active_index()
        if entry is not None:
            runtime = get_company_profile_store().load_active_profile()
            if runtime is None:
                raise ProfileLoadError("COMPANY_PROFILE_LOAD_FAILED")
    except ProfileLoadError as exc:
        raise HTTPException(status_code=503, detail={"fail_class": "COMPANY_PROFILE_LOAD_FAILED"}) from exc
    if entry is None:
        return {
            "schema_version": "company_profile.status.v1",
            "active": False,
            "profile_id": None,
            "profile_digest": None,
            "display_hints": {},
            "raw_text_logged": False,
            "external_send_zero": True,
        }
    return {
        "schema_version": "company_profile.status.v1",
        "active": entry.status == "ACTIVE",
        "profile_id": entry.profile_id,
        "profile_digest": entry.profile_digest,
        "display_hints": entry.display_hints,
        "raw_text_logged": False,
        "external_send_zero": True,
    }


@router.post("/v1/company-profile/register")
async def register_company_profile(
    payload: CompanyProfileRegisterRequest,
    x_admin_role: Optional[str] = Header(default=None),
    x_admin_id_digest: Optional[str] = Header(default=None),
    x_admin_session_digest: Optional[str] = Header(default=None),
    x_admin_auth_method: Optional[str] = Header(default=None),
) -> dict[str, Any]:
    try:
        admin = _admin_context(
            x_admin_role,
            x_admin_id_digest,
            x_admin_session_digest,
            x_admin_auth_method,
        )
        verify_admin_context(
            admin,
            operation="register_company_profile",
        )
        runtime = make_runtime_profile(
            own_bank_holder_aliases=payload.own_bank_holder_aliases,
            own_company_aliases=payload.own_company_aliases,
            own_account_numbers=payload.own_account_numbers,
            own_bank_accounts=[item.model_dump() for item in payload.own_bank_accounts],
        )
        registry_digest = None
        if runtime.own_bank_accounts:
            from butler_pc_core.accounting.assignment.runtime import (
                get_accounting_review_runtime,
            )
            from butler_pc_core.accounting.classify.reconciliation_v2 import tenant_uuid

            registry = get_accounting_review_runtime().store.replace_a4_own_account_registry(
                tenant_id=tenant_uuid(runtime.profile_id),
                accounts=runtime.own_bank_accounts,
                approval_id=admin.admin_id_digest,
                observed_at=datetime.now(timezone.utc)
                .isoformat()
                .replace("+00:00", "Z"),
            )
            registry_digest = registry["registry_digest"]
        entry = get_company_profile_store().save_profile(runtime)
    except AdminAuthError as exc:
        raise HTTPException(status_code=403, detail=admin_error_payload(exc)) from exc
    except ContractValidationError as exc:
        raise HTTPException(status_code=422, detail={"fail_class": str(exc)}) from exc
    except AssignmentError as exc:
        raise HTTPException(status_code=exc.status, detail={"fail_class": exc.code}) from exc
    return {
        "schema_version": "company_profile.register.response.v1",
        "profile_id": entry.profile_id,
        "profile_digest": entry.profile_digest,
        "display_hints": entry.display_hints,
        "raw_text_logged": False,
        "plaintext_persisted": False,
        "external_send_zero": True,
        "a4_registry_ready": bool(runtime.own_bank_accounts),
        "a4_registry_digest": registry_digest,
    }


def _decode_upload_text(raw: bytes) -> str:
    for encoding in ("utf-8-sig", "utf-8", "cp949"):
        try:
            return raw.decode(encoding)
        except UnicodeDecodeError:
            continue
    return raw.decode("utf-8", errors="ignore")


def _extract_holder_candidates(text: str) -> list[str]:
    reader = csv.DictReader(StringIO(text))
    candidate_columns = ("상대계좌예금주명", "보낸분/받는분", "받는분", "보내는분", "거래처", "상대처")
    seen: set[str] = set()
    values: list[str] = []
    for row in reader:
        for column in candidate_columns:
            value = str(row.get(column) or "").strip()
            key = normalize_holder_alias(value)
            if not key or key in seen:
                continue
            seen.add(key)
            values.append(value)
    return values[:10]


@router.post("/v1/company-profile/suggest-from-accounting-file")
async def suggest_from_accounting_file(file: Optional[UploadFile] = File(default=None)) -> dict[str, Any]:
    if file is None:
        raise HTTPException(status_code=422, detail={"fail_class": "COMPANY_PROFILE_FILE_REQUIRED"})
    raw = await file.read()
    text = _decode_upload_text(raw)
    candidates = _extract_holder_candidates(text)
    return {
        "schema_version": "company_profile.suggest.response.v1",
        "suggestions": [
            {
                "value_runtime_text": value,
                "normalized_key_digest": sha256_text(normalize_holder_alias(value)),
                "display_hint": holder_alias_hint(value),
            }
            for value in candidates
        ],
        "candidate_count": len(candidates),
        "persisted": False,
        "plaintext_persisted": False,
        "raw_text_logged": False,
        "external_send_zero": True,
    }

File: butler-desktop/src/lib/companyProfile.ts
import { buildAdminHeaderRecord } from './admin_policy/client';
import type { AdminHeaders, Sha256Digest } from './admin_policy/contracts';
import { sidecarFetch } from './sidecarFetch';

export type CompanyProfileRegisterRequest = Readonly<{
  own_bank_holder_aliases: string[];
  own_company_aliases: string[];
  own_account_numbers: string[];
  own_bank_accounts: [];
}>;

export type CompanyProfileRegisterResponse = Readonly<{
  schemaVersion: 'company_profile.register.response.v1';
  profileId: string;
  profileDigest: Sha256Digest;
  displayHints: Readonly<Record<string, unknown>>;
  rawTextLogged: false;
  plaintextPersisted: false;
  externalSendZero: true;
  a4RegistryReady: boolean;
  a4RegistryDigest: Sha256Digest | null;
}>;

export type CompanyProfileErrorCode =
  | 'COMPANY_PROFILE_HTTP_ERROR'
  | 'COMPANY_PROFILE_CONTENT_TYPE_INVALID'
  | 'COMPANY_PROFILE_RESPONSE_INVALID'
  | 'COMPANY_PROFILE_REQUEST_INVALID';

const RESPONSE_KEYS = [
  'schema_version',
  'profile_id',
  'profile_digest',
  'display_hints',
  'raw_text_logged',
  'plaintext_persisted',
  'external_send_zero',
  'a4_registry_ready',
  'a4_registry_digest',
] as const;
const PROFILE_ID = /^[A-Za-z0-9._:-]{1,128}$/;
const SHA256 = /^sha256:[0-9a-f]{64}$/;

export class CompanyProfileError extends Error {
  constructor(readonly code: CompanyProfileErrorCode) {
    super(code);
    this.name = 'CompanyProfileError';
  }
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === 'object' && value !== null && !Array.isArray(value);
}

function validRuntimeValues(values: string[], required: boolean): boolean {
  return (
    Array.isArray(values)
    && (!required || values.length > 0)
    && values.length <= 50
    && values.every(value => (
      typeof value === 'string'
      && value.length >= 1
      && value.length <= 128
      && value === value.trim()
    ))
  );
}

function validateRequest(request: CompanyProfileRegisterRequest): void {
  if (
    !validRuntimeValues(request.own_bank_holder_aliases, true)
    || !validRuntimeValues(request.own_company_aliases, true)
    || !validRuntimeValues(request.own_account_numbers, false)
    || !Array.isArray(request.own_bank_accounts)
    || request.own_bank_accounts.length !== 0
  ) {
    throw new CompanyProfileError('COMPANY_PROFILE_REQUEST_INVALID');
  }
}

export function parseCompanyProfileRegistration(
  value: unknown,
): CompanyProfileRegisterResponse {
  if (!isRecord(value)) {
    throw new CompanyProfileError('COMPANY_PROFILE_RESPONSE_INVALID');
  }
  const keys = Object.keys(value);
  const registryDigest = value.a4_registry_digest;
  if (
    keys.length !== RESPONSE_KEYS.length
    || RESPONSE_KEYS.some(key => !(key in value))
    || value.schema_version !== 'company_profile.register.response.v1'
    || typeof value.profile_id !== 'string'
    || !PROFILE_ID.test(value.profile_id)
    || typeof value.profile_digest !== 'string'
    || !SHA256.test(value.profile_digest)
    || !isRecord(value.display_hints)
    || value.raw_text_logged !== false
    || value.plaintext_persisted !== false
    || value.external_send_zero !== true
    || typeof value.a4_registry_ready !== 'boolean'
    || (
      registryDigest !== null
      && (typeof registryDigest !== 'string' || !SHA256.test(registryDigest))
    )
  ) {
    throw new CompanyProfileError('COMPANY_PROFILE_RESPONSE_INVALID');
  }
  if (
    (value.a4_registry_ready && registryDigest === null)
    || (!value.a4_registry_ready && registryDigest !== null)
  ) {
    throw new CompanyProfileError('COMPANY_PROFILE_RESPONSE_INVALID');
  }
  return Object.freeze({
    schemaVersion: 'company_profile.register.response.v1',
    profileId: value.profile_id,
    profileDigest: value.profile_digest as Sha256Digest,
    displayHints: Object.freeze({ ...value.display_hints }),
    rawTextLogged: false,
    plaintextPersisted: false,
    externalSendZero: true,
    a4RegistryReady: value.a4_registry_ready,
    a4RegistryDigest: registryDigest as Sha256Digest | null,
  });
}

export async function registerCompanyProfile(
  request: CompanyProfileRegisterRequest,
  admin: AdminHeaders,
): Promise<CompanyProfileRegisterResponse> {
  validateRequest(request);
  const response = await sidecarFetch('/v1/company-profile/register', {
    method: 'POST',
    headers: buildAdminHeaderRecord(admin),
    body: JSON.stringify(request),
  });
  if (!response.ok) {
    throw new CompanyProfileError('COMPANY_PROFILE_HTTP_ERROR');
  }
  const contentType = response.headers.get('content-type') ?? '';
  if (!/^application\/json(?:;|$)/i.test(contentType)) {
    throw new CompanyProfileError('COMPANY_PROFILE_CONTENT_TYPE_INVALID');
  }
  let payload: unknown;
  try {
    payload = await response.json();
  } catch {
    throw new CompanyProfileError('COMPANY_PROFILE_RESPONSE_INVALID');
  }
  return parseCompanyProfileRegistration(payload);
}
